src/linter/root.go

```diff
@@ -1413,40 +1416,45 @@ func (d *rootWalker) parseFuncArgs(params []ir.Node, parTypes phpDocParamsMap, s
 	for _, param := range params {
 		p := param.(*ir.Parameter)
 		v := p.Variable
-		parTyp := parTypes[v.Name]
+		phpDocType := phpDocParamsTypes[v.Name]
 
-		if !parTyp.typ.IsEmpty() {
-			sc.AddVarName(v.Name, parTyp.typ, "param", meta.VarAlwaysDefined)
+		if !phpDocType.typ.IsEmpty() {
+			sc.AddVarName(v.Name, phpDocType.typ, "param", meta.VarAlwaysDefined)
 		}
 
-		typ := parTyp.typ
+		paramTyp := phpDocType.typ
 
-		if p.DefaultValue == nil && !parTyp.optional && !p.Variadic {
+		if p.DefaultValue == nil && !phpDocType.optional && !p.Variadic {
 			minArgs++
 		}
 
 		if p.VariableType != nil {
-			if varTyp, ok := d.parseTypeNode(p.VariableType); ok {
-				typ = varTyp
+			typeHintType, has := d.parseTypeNode(p.VariableType)
+			if has {
+				if !isClosure && !d.typeHintHasMoreAccurateType(typeHintType, phpDocType.typ) {
+					d.Report(p, LevelNotice, "typeHint", "specify the type for the parameter $%s in phpdoc, 'array' type hint is not precise enough", p.Variable.Name)
```

Comment:
I know it's easier to code it this way, but please do not `Report()` from parsing-related functions.
We should separate processing and linting as much as we can.
It makes refactoring impossible when `Report()` could be located anywhere.
Let's not make the situation worse.

Reply:
Done.